src/etos_lib/kubernetes/etos.py

```diff
@@ -0,0 +1,171 @@
+
+    client: DynamicResource
+    namespace: str = "default"
+    strict: bool = False
+
+    @contextmanager
+    def _catch_errors_if_not_strict(self):
+        """Catch errors if the strict flag is False, else raise."""
+        try:
+            yield
+        # Internal exceptions
+        except NoNamespace:
+            if self.strict:
+                raise
+        # Client exceptions
+        except ApiException:
+            if self.strict:
+                raise
+        # Dynamic client exceptions
+        except (ResourceNotFoundError, ResourceNotUniqueError):
+            if self.strict:
+                raise
+        # Built-in exceptions
+        except AttributeError:
+            # AttributeError happens if ResourceNotFoundError was raised when setting up
+            # clients.
+            if self.strict:
+                raise
+
+    def get(self, name: str) -> Optional[ResourceInstance]:
+        """Get a resource from Kubernetes by name."""
+        try:
+            with self._catch_errors_if_not_strict():
+                resource = self.client.get(name=name, namespace=self.namespace)  # type: ignore
+        except NotFoundError:
+            resource = None
+        return resource
+
+    def delete(self, name: str) -> bool:
+        """Delete a resource by name."""
+        with self._catch_errors_if_not_strict():
+            if self.client.delete(name=name, namespace=self.namespace):  # type: ignore
+                return True
+            return False
+
+    def create(self, model: BaseModel) -> bool:
+        """Create a resource from a pydantic model."""
+        with self._catch_errors_if_not_strict():
+            if self.client.create(
+                body=model.model_dump(), namespace=self.namespace
+            ):  # type: ignore
+                return True
+            return False
+
+    def exists(self, name: str) -> bool:
+        """Test if a resource with name exists."""
+        return self.get(name) is not None
+
+
+class Kubernetes:
+    """Kubernetes is a client for fetching ETOS custom resources from Kubernetes."""
+
+    __providers = None
+    __requests = None
+    __testruns = None
+    __environments = None
+    __namespace = None
+    logger = logging.getLogger(__name__)
+
+    def __init__(self, version="v1alpha1"):
+        """Initialize a dynamic client with version."""
+        self.version = f"etos.eiffel-community.github.io/{version}"
+        self.__client = DynamicClient(api_client.ApiClient())
+
+    @property
+    def namespace(self) -> str:
+        """Namespace returns the current namespace of the machine this code is running on."""
+        if self.__namespace is None:
+            if not NAMESPACE_FILE.exists():
+                self.logger.warning(
+                    "Not running in Kubernetes? Namespace file not found: %s", NAMESPACE_FILE
+                )
+                etos_ns = os.getenv("ETOS_NAMESPACE")
+                if etos_ns:
```

Comment:
Are there cases where we want to run this outside kubernetes, but still use the namespace information? Perhaps when testing?

Reply:
Just set the ETOS_NAMESPACE variable in that case?